Fix for issue #960

krmahadevan opened the pull request:
TestNG created the testng-failed.xml all the time 
whereas it should have created this file only when
there are no failures.

Fixed this problem by excluding copying of XmlTest
objects in the deep copy operation of XmlSuite.
Added a unit test so that this is now included as
part of the testing.

krmahadevan commented:
@cedricrefresh @cbeust 
(Not sure which is your id cedric )
Please help with getting this merged.

cbeust commented:
Thanks for the PR, @krmahadevan 

However, this just reverts the commit that caused the bug, which fixes the problem but I thought maybe we could both fix the bug and preserve this commit, which fixed a bug in the cloning of the XML objects.

Any thoughts about that? If not, I'll just go ahead and merge your PR and we'll ping the author of the other commit.

krmahadevan commented:
@cbeust 
As always, you are right :)
I completely missed that part. I should have tried taking into consideration that one as well.  Please give me some time. Let me see if I can fix it such that we aren't reverting the earlier change.

cbeust commented:
Awesome, thanks Krishnan!

krmahadevan commented:
@cbeust 
I amended my previous commit and pushed (forcibly) in the fix. Can you please take a look at this now and see if this is fine with you ? My newly added unit tests are passing with this change.

krmahadevan commented:
Yes I used ```&&``` because we would like to go ahead and check the failedSuite for XmlTests if and only if there are any failures, because now by default failedSuite will always contain XmlTests in it and hence its never going to be empty (due to the deep copy). So I used an ```&&``` operator here.

krmahadevan commented:
But something tells me that this fix may not be correct cedric. Because even though we conditionally generate the testng-failed.xml only if there were failures, we aren't doing anything to remove those already present ```XmlTest``` objects which got added to failedSuite object by default due to the deep copy (from the clone). I think this needs more work :(

So I think this SHOULD NOT be merged yet.

Will try and take another look at this maybe in the evening. Got to leave for work now.

krmahadevan commented:
The only other option I can think of here is to basically have a copy constructor kind of thing in ```XmlSuite``` which doesn't do a deep copy (like the ```clone()```) method and the ```FailedReporter``` resorting to calling the copy constructor instead of using the ```clone()``` method. That would also be a possible solution. What do you think of this option @cbeust  ?

juherr commented on src/test/java/test/failedreporter/FailedReporterLocalTestClass.java:
```diff
@@ -0,0 +1,27 @@
+    public static class WithFailure {
+        @Test
+        public void testMethod() {
+            Assert.assertTrue(false);
```
I think `Assert.fail()` is easier to understand.

juherr commented on src/test/java/test/failedreporter/FailedReporterScenariosTest.java:
```diff
@@ -0,0 +1,81 @@
+            }
+        }
+    }
+
```
I think a test case is missing: having `WithoutFailure` AND `WithFailure` in the suite and check if `TESTNG_FAILED_XML` has only `WithFailure` into it.

juherr commented on src/test/java/test/failedreporter/FailedReporterScenariosTest.java:
```diff
@@ -0,0 +1,81 @@
+        return fileLocation;
+    }
+
+    private static void deleteRecursive(File path) {
```
As TestNG is now using Java7, we can use the nio api: https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#deleteIfExists(java.nio.file.Path)

juherr commented on src/test/java/test/failedreporter/FailedReporterScenariosTest.java:
```diff
@@ -0,0 +1,81 @@
+
+    private File runTests(boolean simulateFailures) {
+        String suiteName = UUID.randomUUID().toString();
+        File fileLocation = new File(tempDir, suiteName);
```
Or http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#createTempDirectory(java.lang.String,%20java.nio.file.attribute.FileAttribute...) with NIO

krmahadevan commented:
@juherr 
Thank you so much for calling out those observations. I will go ahead and close this PR.. work on addressing all the review comments and then raise a new PR.

juherr commented:
Or you can update this one, as you prefer :)

About the fix itself, it was simple, then good for me, but I wonder if `FailedReporter` is not a bit too complicate for the job it is supposed to do (list all failed tests)?
